### PR TITLE
bpo-46025: Fix a crash in the atexit module for auto-unregistering functions

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue46025

### CODE REVIEW

File: Modules/atexitmodule.c [hunk 1]
```diff
@@ -93,13 +93,17 @@ atexit_callfuncs(struct atexit_state *state)
             continue;
         }
 
+        // Increment the refcount of cb->func as the call itself may unregister it
```
Comment (vstinner) [inline Modules/atexitmodule.c]:
```suggestion
        // bpo-46025: Increment the refcount of cb->func as the call itself may unregister it
```

File: Lib/test/_test_atexit.py [hunk 2]
```diff
@@ -116,6 +116,17 @@ def test_bound_methods(self):
...
         self.assertEqual(l, [5])
 
+    def test_atexit_with_unregistered_function(self):
+        # See bpo-46025 for more info
+        def func():
+            atexit.unregister(func)
+            1/0
+        atexit.register(func)
+        try:
+            atexit._run_exitfuncs()
```
Comment (vstinner) [inline Lib/test/_test_atexit.py]: Would it make sense to use support.catch_unraisable_exception() to check if sys.unraisablehook has been called?

File: Modules/atexitmodule.c [hunk 3]
```diff
@@ -93,13 +93,17 @@ atexit_callfuncs(struct atexit_state *state)
...
+        PyObject* the_func = cb->func;
+        Py_INCREF(the_func);
```
Comment (tiran) [inline Modules/atexitmodule.c]:
```suggestion
        PyObject* the_func = Py_NewRef(cb->func);
```

File: Lib/test/_test_atexit.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/_test_atexit.py]: 👍

Comment (pablogsal) [reply Lib/test/_test_atexit.py]: Done!

File: Lib/test/_test_atexit.py [hunk 4]
```diff
@@ -116,6 +116,21 @@ def test_bound_methods(self):
...
+        def func():
+            atexit.unregister(func)
+            1/0
+        atexit.register(func)
+        try:
+            with support.catch_unraisable_exception() as cm:
+                atexit._run_exitfuncs()
+                self.assertEqual(cm.unraisable.object, func)
+                self.assertEqual(cm.unraisable.exc_type, ZeroDivisionError)
+                self.assertEqual(type(cm.unraisable.exc_value), ZeroDivisionError)
```
Comment (vstinner) [inline Lib/test/_test_atexit.py:130]: Not sure if checking exc_value is _needed_, but it's ok to test it :-)

**Review by vstinner — APPROVED** [review]
LGTM. Thanks, I'm feeling more confident with catch_unraisable_exception.

I'm surprised that no other atexit uses catch_unraisable_exception.

File: Lib/test/_test_atexit.py (hunk 4 above)
Comment (pablogsal) [reply Lib/test/_test_atexit.py:130]: This makes it a military-grade test :)

**Comment by vstinner** [comment]
Thanks for the fix ;-)

**Pull request merged**